File: examples/state space Schaefer models/schaefer_r_posfun/schaefer_r.cpp

 //-----------------------------------------------------
 //  State-Space Schaefer Production model
 //  with posfun()
 //
 //  Group project by:
 //  
 //  Yi-Jay Chang
 //  Felipe Carvalho
 //  Jin Gao
 //  Marc Nadon
 //  2/9/2016 9:40:46 PM
 //-----------------------------------------------------

#include <TMB.hpp>

template<class Type>
Type posfun(Type x, Type eps, Type &pen)
{
  if ( x >= eps ){
    return x;
  } else {
    pen += Type(0.01) * pow(x-eps,2);
    return eps/(Type(2.0)-x/eps);
  }
}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(C);
  DATA_VECTOR(I);
  int n = C.size();

  PARAMETER(logR);
  PARAMETER(logK);
  PARAMETER(logQ);
  
  // randon effect
  PARAMETER_VECTOR(logu);
  
  PARAMETER(logSigmaProc); 
  PARAMETER(logSigmaObs);
 
  Type r = exp(logR);
  Type k = exp(logK);
  Type q = exp(logQ);
  
  Type sigmaProc = exp(logSigmaProc);
  Type SigmaObs = exp(logSigmaObs);

  vector<Type> u(n);
  vector<Type> B(n);
  vector<Type> Ihat(n);
  vector<Type> P(n);
 
  for(int t=0; t<n; t++){
  u(t)=exp(logu(t));	  
  }

  Type f = 0.;
  Type fpen = 0.;
  
  // initial condition
  P(0)=u(0)/k;	  

  // process error model
  for(int t=0; t<(n-1); t++){
  B(t+1) = posfun(u(t) + r*u(t)*(Type(1.0)-u(t)/k) - C(t),Type(0.001),fpen); 
  f -= dnorm(log(u[t+1]), log(B(t+1)), sigmaProc, true);
  }
  f += fpen;

  // observation error model
  for(int t=0; t<n; t++){
  Ihat(t) = q*u(t);
  f -= dnorm(log(I(t)), log(Ihat(t)), SigmaObs, true);
  }
  
  // rest of B/k
  for(int t=1; t<n; t++){
  P(t)=u(t)/k;	  
  }
   
  // Penalty on P
  f -= dnorm(P(0), Type(0.8), Type(1), true);
 
      
  ADREPORT(u); // uncertainty  
  ADREPORT(B); // uncertainty
  ADREPORT(Ihat);
    
  REPORT(u);
  REPORT(B);
  REPORT(Ihat);

  REPORT(P);  
  REPORT(sigmaProc);
  REPORT(SigmaObs);
  REPORT(r);
  REPORT(k);
  REPORT(q);
  REPORT(fpen);
     
  return f;
}

File: examples/state space Schaefer models/schaefer_r_posfun/schaefer_r.R
 ################################################
 #  State-Space Schaefer Production model
 #  with posfun()
 #
 #  Group project by:
 #  
 #  Yi-Jay Chang
 #  Felipe Carvalho
 #  Jin Gao
 #  Marc Nadon
 #  2/9/2016 9:40:46 PM
 ################################################
  
 setwd("C:\\Users\\Yi-Jay.Chang\\Desktop\\schaefer_r_posfun\\")
 hake <- read.table("schaefer.dat", header=TRUE)
 names(hake) <- c("t", "C", "I")
 n = length(hake[,1])
 
 # compile and load
 library(TMB)
 compile("schaefer_r.cpp")
 dyn.load(dynlib("schaefer_r"))

 # MakeADFun and output
 Random = c("logu")
 parameters <- list(logR=-1.1, logK=8.0, logQ=-7.9, logu = rep(log(5000),n), logSigmaProc=log(0.1), logSigmaObs=log(0.1))
 model <- MakeADFun(data=hake,parameters=parameters,random=Random)
 model$gr(model$par)
 
 fit <- nlminb(model$par,objective=model$fn, gradient=model$gr,control=list("trace"=1))
 model$gr(fit$par)
 
 rep <- model$report()
 SD = summary(sdreport(model))
 print(rep)
 print(SD)
 
 # Plot
 hake$B <- model$report()$B
 hake$Ihat <- model$report()$Ihat

 par(mfrow=c(1,2))

 plot(hake$t,rep$u,ylim=c(0,5000),type="l",lty=2,col=2,ylab="Catch and Biomass",xlab="Year")
 points(hake$t,hake$C,type="l") 
 points(hake$t,rep$u+1.96*SD[30:53,2],type="l",lty=3,col=2)
 points(hake$t,rep$u-1.96*SD[30:53,2],type="l",lty=3,col=2)

 legend("topright", legend = c("Catch", "Biomass"),
 pch = c(NA, NA), lty = c(1, 2), lwd = c(1,1),
 col = c("black","red"),pt.bg = c(NA,NA))

 plot(I~t, hake, ylim=c(0,1.1*max(hake$I)), ylab="Index",xlab="Year")
 lines(Ihat~t, hake)

